### PR TITLE
[TypeChecker] SE-0213: Implement literal init via coercion

### PR DESCRIPTION
**Pull request by xedin**
Implementation is as follows: In `preCheckExpression` try to
detect if there is `T(literal)` call in the AST, replace it with
implicit `literal as T`, while trying to form type-checked AST,
after constraint solving, restore source information and drop unnecessary
coercion expression.

Resolves: rdar://problem/17088188
Resolves: rdar://problem/39120081
Resolves: rdar://problem/23672697
Resolves: rdar://problem/40379985

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please benchmark

File: test/SILOptimizer/diagnostic_constant_propagation_floats_x86.swift [hunk 1]
```diff
@@ -109,8 +109,5 @@ func testIntToFloatConversion() {
   let e2: Float80 =  18_446_744_073_709_551_617 // expected-warning {{'18446744073709551617' is not exactly representable as 'Float80'; it becomes '18446744073709551616'}}
   _blackHole(e2)
 
-  // No warnings are emitted for conversion through explicit constructor calls.
-  // Note that the error here is because of an implicit conversion of the input
-  // literal to 'Int'.
-  _blackHole(Float80(18_446_744_073_709_551_617)) // expected-error {{integer literal '18446744073709551617' overflows when stored into 'Int'}}
+  _blackHole(Float80(18_446_744_073_709_551_617)) // Ok
```
Comment (ravikandhadai) [inline test/SILOptimizer/diagnostic_constant_propagation_floats_x86.swift]: @xedin Nice to see that the above FIXME's are addressed by this fix. Thanks!

File: test/Sema/diag_erroneous_iuo.swift [hunk 2]
```diff
@@ -116,9 +116,9 @@ _ = Array<Int!>() // expected-error {{'!' is not allowed here; perhaps '?' was i
 let _: Array<Int!> = [1] // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{17-18=?}}
 _ = [Int!]() // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{9-10=?}}
 let _: [Int!] = [1] // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{12-13=?}}
-_ = Optional<Int!>(nil) // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{17-18=?}}
+_ = Optional<Int!>(nil) // expected-error 2 {{'!' is not allowed here; perhaps '?' was intended?}}{{17-18=?}}
 let _: Optional<Int!> = nil // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{20-21=?}}
-_ = Int!?(0) // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{8-9=?}}
+_ = Int!?(0) // expected-error 3 {{'!' is not allowed here; perhaps '?' was intended?}}{{8-9=?}}
```
Comment (jrose-apple) [inline test/Sema/diag_erroneous_iuo.swift]: This looks like we're processing the type multiple times…

Comment (xedin) [reply test/Sema/diag_erroneous_iuo.swift]: Yeah, we have that in a couple of places already unfortunately, because we don't track if we produced the same diagnostic multiple times...

Comment (slavapestov) [reply test/Sema/diag_erroneous_iuo.swift]: Looking at the TypeLoc, we should see it has already been resolved to a Type, and not resolve the TypeRepr again. Where is the diagnostic being emitted?

Comment (xedin) [reply test/Sema/diag_erroneous_iuo.swift]: Ah I see, it's probably my calls to resolveType which caused this, I'll try to fix it.

Comment (xedin) [reply test/Sema/diag_erroneous_iuo.swift]: Done, just needed to cache the type in a couple of places :)

**Comment by xedin** [comment]
@swift-ci please benchmark

**Comment by xedin** [comment]
@swift-ci please benchmark

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Review by rudkx — APPROVED** [review]
LGTM, thanks!

**Review by moiseev — APPROVED** [review]
stdlib change looks innocent. 👍

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
Oops, looks like I missed one 32-bit test.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckConstraints.cpp [hunk 3]
```diff
@@ -1797,6 +1804,64 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
...
+    options |= TypeResolutionFlags::InExpression;
+    type = TC.resolveType(rep, DC, options);
+    typeExpr->getTypeLoc().setType(type);
+  }
+
+  if (!type)
+    return nullptr;
+
+  // Don't bother to convert deprecated selector syntax.
+  if (auto selectorTy = TC.getObjCSelectorType(DC)) {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckConstraints.cpp]: This is the use of string literals for selectors? I wonder if we can (separately) kill that now...

File: stdlib/public/core/Integers.swift.gyb [hunk 4]
```diff
@@ -3590,10 +3590,12 @@ ${assignmentOperatorComment(x.operator, True)}
   /// to find an absolute value. In addition, because `abs(_:)` always returns
   /// a value of the same type, even in a generic context, using the function
   /// instead of the `magnitude` property is encouraged.
-  @_transparent
   public var magnitude: U${Self} {
-    let base = U${Self}(_value)
-    return self < (0 as ${Self}) ? ~base + 1 : base
+    @inline(__always)
+    get {
```
Comment (DougGregor) [inline stdlib/public/core/Integers.swift.gyb]: This seems unrelated?

Comment (xedin) [reply stdlib/public/core/Integers.swift.gyb]: This is needed because constant propagation now looks at both branches, so as a way to work around that behavior @ravikandhadai and @moiseev told me that it would be appropriate to switch from `@_transparent` to `@inlineable`.

Comment (moiseev) [reply stdlib/public/core/Integers.swift.gyb]: It is, unfortunately. `@_transparent` here causes the compile-time overflow on a branch that would not be taken at runtime.

File: lib/Sema/TypeCheckConstraints.cpp [hunk 5]
```diff
@@ -1797,6 +1804,64 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
...
 }
 
+Expr *PreCheckExpression::simplifyTypeConstructionWithLiteralArg(Expr *E) {
+  // If constructor call is expected to produce an optional let's not attempt
+  // this optimization because literal initializers aren't failable.
+  if (!ExprStack.empty()) {
+    auto *parent = ExprStack.back();
+    if (isa<BindOptionalExpr>(parent) || isa<ForceValueExpr>(parent))
+      return nullptr;
+  }
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/TypeCheckConstraints.cpp]: I was just trying to avoid breaking existing code like https://github.com/apple/swift/pull/17860/files#diff-de58b3c2866a24533bc39d1bfd40fc3cR276 but maybe I shouldn't do it at all then?

File: lib/Sema/TypeCheckConstraints.cpp (hunk 3 above)
Comment (xedin) [reply lib/Sema/TypeCheckConstraints.cpp]: That's right, it's `#selector(...)` vs. `Selector("...")`

File: stdlib/public/core/Integers.swift.gyb (hunk 4 above)
Comment (ravikandhadai) [reply stdlib/public/core/Integers.swift.gyb]: Just to clarify. The changes to how initializers are resolved, as a side-effect, makes constant propagation more precise as it can now propagate through more initializers. While in general it is a good thing, in this specific case, it resulted in a false positive due to mandatory inlining.

File: lib/Sema/TypeCheckConstraints.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/TypeCheckConstraints.cpp]: Alright, this check is hidden under `!isSwiftVersionAtLeast(5)` now.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by DougGregor — APPROVED** [review]
LGTM, thanks!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: stdlib/public/core/Integers.swift.gyb (hunk 4 above)
Comment (moiseev) [reply stdlib/public/core/Integers.swift.gyb]: It is very much possible (and probably reasonable) to try the `&+` approach in a separate PR once this lands.

Comment (ravikandhadai) [reply stdlib/public/core/Integers.swift.gyb]: @moiseev @xwu there is one other issue with "magnitude" (besides the +) that could trigger static errors if it is made @_transparent. It is passing a potentially negative (builtin Int) into a UInt : `let base = U${Self}(_value)`. This could trigger a static error on negative values like (-1).magnitude etc.  So perhaps keeping it @inline is better.

Comment (xwu) [reply stdlib/public/core/Integers.swift.gyb]: No, `_value` is [the same type](https://github.com/xedin/swift/blob/29bbc99780265b6800f917f50f24f6818e9ae8e9/stdlib/public/core/Integers.swift.gyb#L3710) regardless of whether it's `Int` or `UInt`, and that initializer simply assigns `self._value = other._value`.
